include/spi.h

```diff
@@ -27,173 +37,184 @@
+ */
+#define SPI_WORD_SIZE_MASK	(0x3F << 5)
+#define SPI_WORD_SIZE_GET(_operation_)		\
+	(((_operation_) & SPI_WORD_SIZE_MASK) >> 5)
+
+#define SPI_WORD(_word_size_)			\
+	((_word_size_) << 5)
 
 /**
- * @brief SPI configuration structure.
- *
- * config is a bit field with the following parts:
- *    mode           [ 0 : 2 ]   - Polarity, phase and loop mode.
- *    transfer_mode  [ 3 ]       - LSB or MSB first transfer mode.
- *    word_size      [ 4 : 11 ]  - Size of a data frame in bits.
- *    RESERVED       [ 12 : 31 ] - Undefined or device-specific usage.
- *
- * max_sys_freq is the clock divider supported by the the host
- * spi controller.
+ * @brief SPI MISO lines
  */
-struct spi_config {
-	u32_t	config;
-	u32_t	max_sys_freq;
-};
+#define SPI_LINES_SINGLE	(0)
+#define SPI_LINES_DUAL		BIT(11)
+#define SPI_LINES_QUAD		BIT(12)
 
-struct spi_buf {
+#define SPI_LINES_MASK		(0x3 << 11)
 
+/**
+ * @brief SPI Chip Select control structure
+ *
+ * gpio_dev is a valid pointer to an actual GPIO device
+ * gpio_pin is a number representing the gpio PIN that will be used
+ *    to act as a CS line
+ * delay is a delay in microseconds to wait before starting the
+ *    transmission and before releasing the CS line
+ */
+struct spi_cs_control {
+	struct device	*gpio_dev;
+	u32_t		gpio_pin;
+	u32_t		delay;
 };
 
 /**
- * @typedef spi_api_configure
- * @brief Callback API upon configuring the const controller
- * See spi_configure() for argument description
+ * @brief SPI controller configuration structure
+ *
+ * frequency is the bus frequency in Hertz
+ * operation is a bit field with the following parts:
+ *    operational mode    [ 0 ]       - master or slave.
+ *    mode                [ 1 : 3 ]   - Polarity, phase and loop mode.
+ *    transfer            [ 4 ]       - LSB or MSB first.
+ *    word_size           [ 5 : 10 ]  - Size of a data frame in bits.
+ *    lines               [ 11 : 12 ] - MISO lines: Single/Dual/Quad.
+ *    RESERVED            [ 13 : 15 ] - Unused yet
+ *
+ * slave is the slave number from 0 to host constoller slave limit.
+ *
+ * cs is a valid pointer on a struct spi_cs_control is CS line is
+ *    emulated through a gpio line, or NULL otherwise.
  */
-typedef int (*spi_api_configure)(struct device *dev,
-				 struct spi_config *config);
+struct spi_config {
+	u32_t		frequency;
+	u16_t		operation;
+	u16_t		slave;
+	struct spi_cs_control *cs;
+};
+
 /**
- * @typedef spi_api_slave_select
- * @brief Callback API upon selecting a slave
- * See spi_slave_select() for argument description
+ * @brief SPI buffer structure
+ *
+ * buf is a valid pointer on a data buffer, or NULL otherwise.
+ * len is the length of the buffer or, if buf is NULL, will be the
+ *     length which as to be sent as dummy bytes (as TX buffer) or
+ *     the length of bytes that should be skipped (as RX buffer).
  */
-typedef int (*spi_api_slave_select)(struct device *dev, u32_t slave);
+struct spi_buf {
+	void *buf;
+	u32_t len;
```

Comment:
indeed, should be the same actually as zephyr currently only support 32bits arch.